psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,257 @@
+from psyneulink.core.components.functions.transferfunctions import Linear, Logistic, ReLU
+from psyneulink.library.compositions.pytorchllvmhelper import *
+from psyneulink.core import llvm as pnlvm
+
+try:
+    import torch
+    from torch import nn
+    torch_available = True
+except ImportError:
+    torch_available = False
+
+
+# helper method that identifies the type of function used by a node, gets the function
+# parameters and uses them to create a function object representing the function, then returns it
+def pytorch_function_creator(function, device, context=None):
+    def get_fct_param_value(param_name):
+        val = function._get_current_function_param(
+            param_name, context=context)
+        if val is None:
+            val = getattr(function.defaults, param_name)
+
+        return float(val)
+
+    if isinstance(function, Linear):
+        slope = get_fct_param_value('slope')
+        intercept = get_fct_param_value('intercept')
+        return lambda x: x * slope + intercept
+
+    elif isinstance(function, Logistic):
+        gain = get_fct_param_value('gain')
+        bias = get_fct_param_value('bias')
+        offset = get_fct_param_value('offset')
+        return lambda x: 1 / (1 + torch.exp(-gain * (x + bias) + offset))
+
+    elif isinstance(function, ReLU):
+        gain = get_fct_param_value('gain')
+        bias = get_fct_param_value('bias')
+        leak = get_fct_param_value('leak')
+        return lambda x: (torch.max(input=(x - bias), other=torch.tensor([0], device=device).double()) * gain +
+                            torch.min(input=(x - bias), other=torch.tensor([0], device=device).double()) * leak)
+
+    else:
+        raise Exception(f"Function {function} is not currently supported in AutodiffCompositions!")
+
+# Helper method that creates a bin func that returns the derivative of the function into the builder
+# FIXME: Add compiled derivative functions, and move these calls there
+def bin_function_derivative_creator(ctx, node, context=None):
+    # first try to get cached func
```

Comment:
The function should be generated from an object rather than a name, caching then works transparently and no of this is needed.

Reply:
trying to implement this by adding a class for the derivative function, but I seem to be getting a segfault due to a parsing error; 
any pointers on how I should approach this?

Reply:
using `PNL_LLVM_DEBUG=llvm` will dump the llvm assembly files (~one per function) you can then check which function generates invalid syntax

Reply:
it might alos be better to just use `node` and select derivative using `tags`